src/python/pants/backend/python/macros/deprecation_fixers.py

```diff
@@ -0,0 +1,221 @@
+
+
+@dataclass(frozen=True, order=True)
+class GeneratorRename:
+    build_path: str
+    alias: str
+    new_name: str | None
+
+
+@dataclass(frozen=True)
+class MacroRenames:
+    generators: tuple[GeneratorRename, ...]
+    # Includes the alias for the macro.
+    generated: FrozenDict[Address, tuple[Address, str]]
+
+
+@rule(desc="Determine how to rename Python macros to target generators", level=LogLevel.DEBUG)
+async def determine_macro_changes(all_targets: AllTargets) -> MacroRenames:
+    # Strategy: Find `python_requirement` targets who depend on a `_python_requirements_file`
+    # target to figure out which macros we have. Note that context-aware object factories (CAOFs)
+    # are not actual targets and are "erased", so this is the way to find the macros.
+    #
+    # We also need to figure out if the new target generator can use the default `name=None` or
+    # if it needs to set an explicit name, based on whether it's the build root and whether the
+    # default is already taken.
+
+    dirs_with_default_name = set()
+    python_requirement_dependencies_fields = set()
+    for tgt in all_targets:
+        if tgt.address.is_default_target:
+            dirs_with_default_name.add(tgt.address.spec_path)
+        if isinstance(tgt, PythonRequirementTarget) and tgt[Dependencies].value is not None:
+            python_requirement_dependencies_fields.add(tgt[Dependencies])
+
+    build_file_addresses_per_tgt = await MultiGet(
+        Get(BuildFileAddress, Address, deps_field.address)
+        for deps_field in python_requirement_dependencies_fields
+    )
+    explicit_deps_per_tgt = await MultiGet(
+        Get(ExplicitlyProvidedDependencies, DependenciesRequest(deps_field))
+        for deps_field in python_requirement_dependencies_fields
+    )
+    deps_per_tgt = await MultiGet(
+        Get(UnexpandedTargets, Addresses(explicit_deps.includes))
+        for explicit_deps in explicit_deps_per_tgt
+    )
+
+    generators = set()
+    generated = {}
+    for python_req_deps_field, build_file_addr, deps in zip(
+        python_requirement_dependencies_fields, build_file_addresses_per_tgt, deps_per_tgt
+    ):
+        generator_tgt = next((tgt for tgt in deps if isinstance(tgt, PythonRequirementsFile)), None)
+        if generator_tgt is None:
+            continue
+
+        # Assume there is a single source. We should have changed this target to use
+        # `SingleSourceField`, alas.
+        generator_source = generator_tgt[MultipleSourcesField].value[0]  # type: ignore[index]
+        if "Pipfile" in generator_source:
+            generator_alias = "pipenv_requirements"
+        elif "pyproject.toml" in generator_source:
+            generator_alias = "poetry_requirements"
+        else:
+            generator_alias = "python_requirements"
+
+        # TODO: Robustly handle if the `name` is already claimed? This can happen, for example,
+        #  if you have two `python_requirements` in the same BUILD file. Perhaps error?
```

Comment:
Any sense if users actually have >1 requirements.txt in the same dir? 

I'd like to still experiment more what happens when you have two `python_requirements` CAOFs in the same directory - I honestly don't know. Plausible we don't even support that.

Reply:
I suspect that it is rare enough not to worry about here.